lib/eigensolve_quda.cpp

```diff
@@ -0,0 +1,948 @@
+                        tmp[0]->Ncolor(), tmp[0]->Nspin(), Nvec, 0, (char **)0);
+
+      host_free(V);
+      if (eig_vecs[0]->Location() == QUDA_CUDA_FIELD_LOCATION) {
+        for (int i = 0; i < Nvec; i++) {
+          *eig_vecs[i] = *tmp[i];
+          delete tmp[i];
+        }
+      }
+
+      if (getVerbosity() >= QUDA_SUMMARIZE) printfQuda("Done loading vectors\n");
+    } else {
+      errorQuda("No eigenspace input file defined.");
+    }
+#else
+    errorQuda("\nQIO library was not built.\n");
+#endif
+    // profile.TPSTOP(QUDA_PROFILE_IO);
+    // profile.TPSTART(QUDA_PROFILE_COMPUTE);
+
+    return;
+  }
+
+  void EigenSolver::saveVectors(const std::vector<ColorSpinorField *> &eig_vecs, std::string vec_outfile)
+  {
+    // profile.TPSTOP(QUDA_PROFILE_COMPUTE);
+    // profile.TPSTART(QUDA_PROFILE_IO);
+
+#ifdef HAVE_QIO
+    const int Nvec = eig_vecs.size();
+    std::vector<ColorSpinorField *> tmp;
+    if (eig_vecs[0]->Location() == QUDA_CUDA_FIELD_LOCATION) {
+      ColorSpinorParam csParam(*eig_vecs[0]);
+      csParam.fieldOrder = QUDA_SPACE_SPIN_COLOR_FIELD_ORDER;
+      csParam.setPrecision(eig_vecs[0]->Precision() < QUDA_SINGLE_PRECISION ? QUDA_SINGLE_PRECISION :
+                                                                              eig_vecs[0]->Precision());
+      csParam.location = QUDA_CPU_FIELD_LOCATION;
+      csParam.create = QUDA_NULL_FIELD_CREATE;
+      for (int i = 0; i < Nvec; i++) {
+        tmp.push_back(ColorSpinorField::Create(csParam));
+        *tmp[i] = *eig_vecs[i];
+      }
+    } else {
+      for (int i = 0; i < Nvec; i++) { tmp.push_back(eig_vecs[i]); }
+    }
+
+    if (getVerbosity() >= QUDA_SUMMARIZE) printfQuda("Start saving %d vectors to %s\n", Nvec, vec_outfile.c_str());
+
+    void **V = static_cast<void **>(safe_malloc(Nvec * sizeof(void *)));
+    for (int i = 0; i < Nvec; i++) {
+      V[i] = tmp[i]->V();
+      if (V[i] == NULL) {
+        if (getVerbosity() >= QUDA_SUMMARIZE) printfQuda("Could not allocate space for eigenVector[%04d]\n", i);
+      }
+    }
+
+    write_spinor_field(vec_outfile.c_str(), &V[0], tmp[0]->Precision(), tmp[0]->X(),
+                       tmp[0]->Ncolor(), tmp[0]->Nspin(), Nvec, 0, (char **)0);
+
+    host_free(V);
+    if (getVerbosity() >= QUDA_SUMMARIZE) printfQuda("Done saving vectors\n");
+    if (eig_vecs[0]->Location() == QUDA_CUDA_FIELD_LOCATION) {
+      for (int i = 0; i < Nvec; i++) delete tmp[i];
+    }
+
+#else
+    errorQuda("\nQIO library was not built.\n");
+#endif
+    // profile.TPSTOP(QUDA_PROFILE_IO);
+    // profile.TPSTART(QUDA_PROFILE_COMPUTE);
+
+    return;
+  }
+
+  void EigenSolver::loadFromFile(const DiracMatrix &mat, std::vector<ColorSpinorField *> &kSpace,
+                                 std::vector<Complex> &evals)
+  {
+
+    if (getVerbosity() >= QUDA_SUMMARIZE) printfQuda("Loading eigenvectors\n");
+    loadVectors(kSpace, eig_param->vec_infile);
```

Comment:
This comment would be better put in `qio_field.cpp`, but anyway: I think we should have a more constructive error when a file doesn't exist, especially that loading/saving eigenvectors is becoming a lot more important in QUDA with all these additions. Right now the only error message in `qio_field.cpp` is that "Open file failed!", and I think we should say something more useful (if QIO gives us that info), i.e., if we can discriminate "file doesn't exist" from "file isn't properly formatted", that's a little thing that goes a long way.

Reply:
Is this a general comment about how our QIO code handles errors, or about how I use it in the eigensolver? Shall we fix the QIO `stdout` on this pull?

Reply:
Ok, while that specific print statement doesn't belong there since there is another equivalent statement in `EigenSolver::loadVectors` which called from this function (and so I'll remove it on my next push), I do see that the info from QIO is actually not bad.  E.g., if QIO fails to find a file, it will print the following
```
QIO_create_reader(0): cannot open conf_stout_1e4.1000__level_0_nvec_6 as SINGLEFILE; trying PARTFILE
QIO_create_reader(0): cannot open conf_stout_1e4.1000__level_0_nvec_6.vol0000 as PARTFILE; trying PARTFILE_DIR
QIO_create_reader(0): cannot open vol0000/conf_stout_1e4.1000__level_0_nvec_6 as PARTFILE_DIR; QIO_create_reader FAILED
MG level 0 (GPU): open_test_input(0): QIO_open_read returns NULL.
MG level 0 (GPU): ERROR: MG level 0 (GPU): ERROR: Open file failed
```
This seems plenty informative.